### PR TITLE
[FrameworkBundle] Allow using the kernel as a registry of controllers and service factories

### LINKED ISSUE #28992: [DependencyInjection] Request for anonymous function factories
It was already proposed few years ago: #12115, but I want to revisit this issue again.

I'm OK if you still don't accept 
```php
return function (ContainerConfigurator $c) {
    $s = $c->services()
        ->defaults()
        ->autoconfigure(true)
        ->autowire(true)
        ->private();

    $s->set('doctrine_pdo', \Doctrine\DBAL\Connection::class)
        ->factory(function () {
            return DriverManager::getConnection(...);
        });

};
```
, but I want it at least for my projects.

Major issues with this proposal were:
1. `use` may hold reference to object and cannot be inlined to the generated static factory.
2. it might be actually _closed by_ (to be a _closure_) some object (`$this`) and unfortunately, PHP documentation makes no difference between anonymous functions and closures. But this one could be solved by requirement to prefix anonymous functions with "static" and throw exception in case of violation:
   ```
    ->factory(
        static function (...) {
            // ...
        }
    )
   ```
3. this one by @stof, but I still don't get it, probably it was about inlining?
  > Btw, I see a big drawback with this approach: any time your factory references another of its methods to get a dependency, it will create its own instance of the services, which will not be the same than the shared service used by the container. 
4. It required to have dependency on `SuperClosure`. 
 
But since you have `symfony/contracts` repo, I think we can add something like `Symfony\Closures\ClosureDumper` interface and make corresponding changes in the `PhpDumper`. It allows you to avoid dependency on 3rd party libraries like `SuperClosure`, but opens extension points for users like me:
```php
namespace Symfony\Component\BoostedReflection;

interface AnonymousFunctionDumper
{
    /**
     * @throws AnonymousFunctionScopeContainsReferences foo () use (&a, $object)
     * @throws AnonymousFunctionIsClosureForReal Holds reference to $this
     *                    i.e. is not an anonymous function, strictly speaking
     * @return PHP code
     */
    public function dump(\Closure $closure): string;
}
```

btw, `Symfony\Component\DependencyInjection\LazyProxy\Instantiator\InstantiatorInterface` could be moved to the `symfony/contracts` too.

So `1.`, `2.` are not problem for me at all; `3.` is weird for me. `4.` not a problem anymore with `symfony/contracts`.

Another possible solution is to allow injecting custom container configurators. I can collect all closures, generate static factories for each of them and replace factory definition with it.

---

Real example: I had a need to integrate Doctrine migrations to the project, but without ORM. I had to translate all [these instructions](https://www.doctrine-project.org/projects/doctrine-migrations/en/1.7/reference/building_framework_specific_integration.html#register-console-commands) into this Symfony fluent PHP dialect of the plain PHP:
```php
    $s->set(Application::class)
        ->call('setDispatcher', [ref(EventDispatcherInterface::class)])
        ->call('setHelperSet', [ref(HelperSet::class)])
        ->call('addCommands', [[
            inline(ExecuteCommand::class),
            inline(GenerateCommand::class),
            inline(LatestCommand::class),
            inline(MigrateCommand::class),
            inline(DiffCommand::class)->args([ref(MigrationsConfigurator::class)]),
            inline(UpToDateCommand::class),
            inline(StatusCommand::class),
            inline(VersionCommand::class)
        ]]);
    $s->set(Symfony\Component\Console\Helper\HelperSet::class)->args([
        [
            'db' => inline(ConnectionHelper::class)->args([ref('doctrine_pdo')]),
            'question' => inline(QuestionHelper::class),
            'configuration' => inline(ConfigurationHelper::class)
                ->args([
                    ref('doctrine_pdo'),
                    inline(Configuration::class)->factory(
                        [ref(MigrationsConfigurator::class), 'createConfiguration']
                    )
                ])
        ]
    ]);
```
instead of copy-pasting:
```php
$s->setApplication::class)
    ->factory(function (PDO $pdo) {
        $cli = new Application();
        $cli->setHelperSet(
            new \Symfony\Component\Console\Helper\HelperSet(array(
                'db' => new \Doctrine\DBAL\Tools\Console\Helper\ConnectionHelper($pdo),
                'question' => new \Symfony\Component\Console\Helper\QuestionHelper(),
            ))
        );
        $cli->addCommands([
            new \Doctrine\DBAL\Migrations\Tools\Console\Command\ExecuteCommand(),
            new \Doctrine\DBAL\Migrations\Tools\Console\Command\GenerateCommand(),
            new \Doctrine\DBAL\Migrations\Tools\Console\Command\LatestCommand(),
            new \Doctrine\DBAL\Migrations\Tools\Console\Command\MigrateCommand(),
            new \Doctrine\DBAL\Migrations\Tools\Console\Command\DiffCommand(
                // blah blah with $pdo
            ),
            new \Doctrine\DBAL\Migrations\Tools\Console\Command\UpToDateCommand(),
            new \Doctrine\DBAL\Migrations\Tools\Console\Command\StatusCommand(),
            new \Doctrine\DBAL\Migrations\Tools\Console\Command\VersionCommand()
        ]);
        return $cli;
    });
```

Yes, the last example is awful, but the last one at least is real PHP. 
Yes, I could create class with static method, but sometimes you're in position like "argh, I'm lazy for both variants, just, just be cool, OK?".

It is possible to add some extension point for this feature?

### LINKED ISSUE #29997: [RFC] [DI] Functional Service Support
A somewhat common pattern seen among the PHP community is making final classes with one method `__invoke` like so:

```php
final class DoSomething {
  public function __constructor(/* args */) {
     // set args
  }
  public function __invoke(): void {
    // do something
  }
}
```

Semantically, this is equivalent to the following done with functions and closures:

```php
function doSomething(/* args */) {
  return function() use (/* args */): void {
    // do something
  };
}
```

I'd like to discuss ways in which we could make the latter (functional service) easier to use in a symfony application.

The former pattern I think emerged in PHP rather than the latter due to PSR-4 composer autoloading defaults and the fact that most autowired di frameworks (like SF DI) use the class name to wire dependencies.

However, using the function/closure definition is much more terse, and allows you to easily include multiple related services into one file.

## Example

With SF's new php fluent container configuration, i've been able to make use of functional services relatively pain free with the help of a few helper functions and abusing the php lexer.

```php
// src/services.php

namespace App;

const serviceA = serviceA::class;
function serviceA(callable $serviceB) { // $serviceB is autowired
   return function() use ($serviceB) {}; // does stuff
}

const serviceB = serviceB::class;
function serviceB(Psr\Log\LoggerInterface $logger) {
     return function(): int {}; // does stuff
}
```

```php
// src/util.php

namespace App;

/**
 * Registers a service that is a function which returns a closure. If a suffix is applied, the function name is suffixed with `.{suffix}`
 */
function fnService(AbstractServiceConfigurator $container, string $fnName, ?string $suffix = null, ?string $bindName = null): ServiceConfigurator {
    $serviceId = $fnName . ($suffix === null ? '' : '.' .$suffix)
    $container->set($serviceId, 'Closure')->factory($fnName);
    if (!$bindName) {
        $parts = explode('\\', $fnName);
        $bindName = end($parts);
    }

    $container->bind($bindName, ref($serviceId);
}
```

```php
// config/services.php

use App;
use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;
use function App\fnService;
use function Symfony\Component\DependencyInjection\Loader\Configurator\ref;

return function($configurator) {
  $container = $configurator->services()
    ->defaults()->autowire()->autoconfigure()->private();

  fnService($container, App\serviceA);
  fnService($container, App\serviceB);
};
```

## Desired Outcomes

I'd be real stoked to see if we could do something like `$container->fn()` which is basically my implementation of `fnService`.

I'm also very open to ideas of how we could possible make this even more robust.

As far as dealing with the pain of having to register non-psr4 files into the `files` section of the composer.json, that can be mitigated with a composer plugin like: https://github.com/krakphp/php-inc.

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #28992, fix #29997
| License       | MIT
| Doc PR        | -

This PR builds on #34873 and #34872 and allows using the `Kernel` as a registry of autowired controllers and service factories. The `ContainerConfigurator` passed to `configureContainer()` defaults to declaring autowired and autoconfigured services.

TL;DR: Silex is back but in a much more powerful way \o/

Here is a Kernel that just works and displays `Hello App\Foo` on the `/` route:
```php
class Kernel extends BaseKernel
{
    use MicroKernelTrait;

    protected function configureContainer(ContainerConfigurator $container): void
    {
        $container->services()
            ->load('App\\', '../src')
            ->set(Foo::class)
                ->factory([$this, 'createFoo']);
    }

    public function createFoo(Bar $bar)
    {
        return new Foo($bar);
    }

    protected function configureRoutes(RoutingConfigurator $routes): void
    {
        $routes->add('home', '/')->controller([$this, 'helloAction']);
    }

    public function helloAction(Foo $foo)
    {
        return new Response('Hello '.get_class($foo));
    }
}
```

### CODE REVIEW

**Review by dunglas — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php [hunk 1]
```diff
@@ -48,29 +36,26 @@ protected function configureRoutes(RouteCollectionBuilder $routes)
      *         ->controller('App\Controller\AdminController::dashboard')
      *     ;
      */
-    protected function configureRouting(RoutingConfigurator $routes): void
-    {
-        @trigger_error(sprintf('Not overriding the "%s()" method is deprecated since Symfony 5.1 and will trigger a fatal error in 6.0.', __METHOD__), E_USER_DEPRECATED);
-    }
+    abstract protected function configureRoutes(RoutingConfigurator $routes);
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php]: Changing the typehint with the existing method name is a BC break.

File: src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php [hunk 2]
```diff
@@ -48,29 +36,26 @@ protected function configureRoutes(RouteCollectionBuilder $routes)
...
-     *     $c->register('halloween', 'FooBundle\HalloweenProvider');
+     *     $c->services()->set('halloween', 'FooBundle\HalloweenProvider');
      *
      * Or parameters:
      *
-     *     $c->setParameter('halloween', 'lot of fun');
+     *     $c->parameters()->set('halloween', 'lot of fun');
      */
-    abstract protected function configureContainer(ContainerBuilder $c, LoaderInterface $loader);
+    abstract protected function configureContainer(ContainerConfigurator $c);
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php]: That's also a BC break. We would need a new trait.

File: src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php [hunk 3]
```diff
@@ -148,6 +167,14 @@ public function loadRoutes(LoaderInterface $loader)
         try {
             $this->configureRoutes(new RoutingConfigurator($collection, $kernelLoader, $file, $file));
 
+            foreach ($collection as $route) {
+                $controller = $route->getDefault('_controller');
+
+                if (($controller[0] ?? null) === $this) {
+                    $route->setDefault('_controller', ['kernel', $controller[1]]);
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php]: What if there is no `$controller[1]` ?

File: src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php [hunk 4]
```diff
@@ -148,6 +167,14 @@ public function loadRoutes(LoaderInterface $loader)
         try {
             $this->configureRoutes(new RoutingConfigurator($collection, $kernelLoader, $file, $file));
 
+            foreach ($collection as $route) {
+                $controller = $route->getDefault('_controller');
+
+                if (($controller[0] ?? null) === $this) {
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php]: What if `$controller` is not an array but a different callable ?

File: src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php]: it's not a BC break because this is a trait: everyone is forced to override it right now.

File: src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php]: it's not a BC break because this is a trait: everyone is forced to override it right now.

File: src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php (hunk 3 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php]: check added

File: src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php (hunk 4 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php]: check added

File: src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php (hunk 2 above)
Comment (stof) [reply src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php]: But isn't it an issue if you override it with a different signature ?

Comment (HeahDude) [reply src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php]: I was surprised at first too, but the trait does not enforce a signature like an abstract class. And this makes sense when you think about it, its definition is just replaced and always has. Good for reusable code not defining contract.

**Comment by nicolas-grekas** [comment]
PR is ready, now with tests.
Failures are all false positives.

**Review by yceruto — APPROVED** [review]
Nice :), LGTM!

File: src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php [hunk 5]
```diff
@@ -99,9 +109,51 @@ public function registerContainerConfiguration(LoaderInterface $loader)
...
+
+                if (0 !== strpos($e->getMessage(), sprintf('Argument 1 passed to %s::configureContainer() must be an instance of %s,', static::class, ContainerConfigurator::class))) {
+                    throw $e;
+                }
+            }
+
+            $defaultDefinition = (new Definition())->setAutowired(true)->setAutoconfigured(true);
+            $kernelLoader = $loader->getResolver()->resolve($file);
+            $kernelLoader->setCurrentDir(\dirname($file));
+            $instanceof = &\Closure::bind(function &() { return $this->instanceof; }, $kernelLoader, $kernelLoader)();
```
Comment (weaverryan) [inline src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php]: Is it possible for there to be any `instanceof` at this point?

File: src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php [hunk 6]
```diff
@@ -95,13 +105,43 @@ public function registerContainerConfiguration(LoaderInterface $loader)
...
+                return;
+            } catch (\TypeError $e) {
+                $file = $e->getFile();
+
+                if (0 !== strpos($e->getMessage(), sprintf('Argument 1 passed to %s::configureContainer() must be an instance of %s,', static::class, ContainerConfigurator::class))) {
+                    throw $e;
+                }
             }
 
-            $this->configureContainer($container, $loader);
```
Comment (weaverryan) [inline src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php]:
```
// the user has opted into using the ContainerConfigurator
```

File: src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php [hunk 7]
```diff
@@ -95,13 +105,43 @@ public function registerContainerConfiguration(LoaderInterface $loader)
...
+            $defaultDefinition = (new Definition())->setAutowired(true)->setAutoconfigured(true);
+            $kernelLoader = $loader->getResolver()->resolve($file);
```
Comment (weaverryan) [inline src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php]:
```
/* @var PhpFileLoader $kernelLoader */
```

it makes the logic a bit easier to follow... 😇

Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php]: added

File: src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php (hunk 6 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php]: added

File: src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php (hunk 5 above)
Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php]: No, but the point here is to get a reference, so it doesn't matter.
Because the property is protected, it is part of the BC promise so we can this this without breaking any contracts.

File: src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php [hunk 8]
```diff
@@ -139,13 +153,22 @@ public function registerContainerConfiguration(LoaderInterface $loader)
     public function loadRoutes(LoaderInterface $loader)
     {
         $file = (new \ReflectionObject($this))->getFileName();
+        /* @var RoutingPhpFileLoader $kernelLoader */
```
Comment (ro0NL) [inline src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php]: `/**`

Comment (nicolas-grekas) [reply src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php]: the codebase uses both styles, that's doesn't make any diff :)

**Review by weaverryan — APPROVED** [review]
I like this = +1

Just make sure you make fabbot happy ;)

**Comment by mnapoli** [comment]
That sounds really interesting! Since you closed #28992 is it possible to use anonymous functions to declare services?

**Comment by nicolas-grekas** [comment]
Anonymous functions won't happen, I've explained why in #28992 (code relocation is full of edge cases we don't want to deal with). But this PR provides an alternative to using closures, by using methods on the Kernel. Let's see how far this goes :)

**Comment by mnapoli** [comment]
Oh OK, thanks for the clarification.

Since #28992 was specifically about closures, and it was marked as "fixed" by this PR, this was confusing.

**Comment by nicolas-grekas** [comment]
Thanks for asking, it deserved the clarification :)
